README.md
New IO for Ruby
===============
[![Build Status](http://travis-ci.org/tarcieri/nio4r.png)](http://travis-ci.org/tarcieri/nio4r)

When it comes to managing many IO objects on Ruby, there aren't a whole lot of
options. The most powerful API Ruby itself gives you is Kernel.select, and
select is hurting when it comes to performance and in terms of having a nice
API.

Once upon a time Java was a similar mess. They got out of it by adding the
Java NIO API. Java NIO provides a high performance selector API for monitoring
large numbers of file descriptors.

This library aims to incorporate the ideas of Java NIO in Ruby. These are:

* Expose high level interfaces for doing high performance IO, but keep the
  codebase small to encourage multiple implementations on different platforms
* Be as portable as possible, in this case across several Ruby VMs
* Provide inherently thread-safe facilities for working with IO objects

Supported Platforms
-------------------

nio4r is known to work on the following Ruby implementations:

* MRI/YARV 1.8.7, 1.9.2, 1.9.3
* REE (2011.12)
* JRuby 1.6.x (and likely earlier versions too)
* Rubinius 1.x/2.0
* A pure Ruby implementation based on Kernel.select is also provided

Platform notes:

* MRI/YARV and Rubinius implement nio4j with a C extension based on libev,
  which provides a high performance binding to native IO APIs
* JRuby uses a Java extension based on the high performance Java NIO subsystem
* A pure Ruby implementation is also provided for Ruby implementations which
  don't implement the MRI C extension API

Usage
-----

### Selectors

The NIO::Selector class is the main API provided by nio4r. Use it where you
might otherwise use Kernel.select, but want to monitor the same set of IO
objects across multiple select calls rather than having to reregister them
every single time:

```ruby
require 'nio'

selector = NIO::Selector.new
```

To monitor IO objects, attach them to the selector with the NIO::Selector#register
method, monitoring them for read readiness with the :r parameter, write
readiness with the :w parameter, or both with :rw.

```ruby
>> reader, writer = IO.pipe
 => [#<IO:0xf30>, #<IO:0xf34>]
>> monitor = selector.register(reader, :r)
 => #<NIO::Monitor:0xfbc>
```

After registering an IO object with the selector, you'll get a NIO::Monitor
object which you can use for managing how a particular IO object is being
monitored. Monitors will store an arbitrary value of your choice, which
provides an easy way to implement callbacks:

```ruby
>> monitor = selector.register(reader, :r)
 => #<NIO::Monitor:0xfbc>
>> monitor.value = proc { puts "Got some data: #{monitor.io.read_nonblock(4096)}" }
 => #<Proc:0x1000@(irb):4>
```

The main method of importance is NIO::Selector#select, which monitors all
registered IO objects and returns an array of monitors that are ready.

```ruby
>> writer << "Hi there!"
 => #<IO:0x103c>
>> ready = selector.select
 => [#<NIO::Monitor:0xfbc>]
>> ready.each { |m| m.value.call }
Got some data: Hi there!
 => [#<NIO::Monitor:0xfbc>]
```

By default, NIO::Selector#select will block indefinitely until one of the IO
objects being monitored becomes ready. However, you can also pass a timeout to
wait in seconds to NIO::Selector#select just like you can with Kernel.select:

```ruby
ready = selector.select(15) # Wait 15 seconds
```

If a timeout occurs, ready will be nil.

You can avoid allocating an array each time you call NIO::Selector#select by
passing a block to select. The block will be called for each ready monitor
object, with that object passed as an argument. The number of ready monitors
is returned as a Fixnum:

```ruby
>> selector.select { |m| m.value.call }
Got some data: Hi there!
 => 1
```

When you're done monitoring a particular IO object, just deregister it from
the selector:

```ruby
selector.deregister(reader)
```

### Monitors

Monitors provide methods which let you introspect on why a particular IO
object was selected. These methods are not thread safe unless you are holding
the selector lock (i.e. if you're in a block pased to #select). Only use them
if you aren't concerned with thread safety, or you're within a #select
block:

- ***#interests***: what this monitor is interested in (:r, :w, or :rw)
- ***#readiness***: what the monitored IO object is ready for according to the last select operation
- ***#readable?***: was the IO readable last time it was selected?
- ***#writable?***: was the IO writable last time it was selected?

Monitors also support a ***#value*** and ***#value=*** method for storing a
handle to an arbitrary object of your choice (e.g. a proc)

### Buffers

NIO::Buffer is a fast byte queue which is primarily intended for non-blocking
I/O applications but is suitable wherever buffering is required.

NIO::Buffer provides a subset of the methods available in Strings:

```ruby
>> buf = NIO::Buffer.new
=> #<NIO::Buffer:0x12fc708>
>> buf << "foo"
=> "foo"
>> buf << "bar"
=> "bar"
>> buf.to_str
=> "foobar"
>> buf.size
=> 6
```

The NIO::Buffer#<< method is an alias for NIO::Buffer#append.  A prepend method
is also available:

```ruby
>> buf = NIO::Buffer.new
=> #<NIO::Buffer:0x12f5250>
>> buf.append "bar"
=> "bar"
>> buf.prepend "foo"
=> "foo"
>> buf.append "baz"
=> "baz"
>> buf.to_str
=> "foobarbaz"
```

NIO::Buffer#read can be used to retrieve the contents of a buffer.  You can mix
reads alongside adding more data to the buffer:

```ruby
>> buf = NIO::Buffer.new
=> #<NIO::Buffer:0x12fc5f0>
>> buf << "foo"
=> "foo"
>> buf.read 2
=> "fo"
>> buf << "bar"
=> "bar"
>> buf.read 2
=> "ob"
>> buf << "baz"
=> "baz"
>> buf.read 3
=> "arb"
```

Finally, NIO::Buffer provides methods for performing non-blocking I/O with the
contents of the buffer. The NIO::Buffer#read_from(IO) method will read as much
data as possible from the given IO object until the read would block.

The NIO::Buffer#write_to(IO) method writes the contents of the buffer to the
given IO object until either the buffer is empty or the write would block:

```ruby
>> buf = NIO::Buffer.new
=> #<NIO::Buffer:0x12fc5c8>
>> file = File.open("README")
=> #<File:README>
>> buf.read_from(file)
=> 1713
>> buf.to_str
=> "= NIO::Buffer\n\nNIO::Buffer is a fast byte queue...
```

If the file descriptor is not ready for I/O, the Errno::EAGAIN exception is
raised indicating no I/O was performed.

Concurrency
-----------

nio4r provides internal locking to ensure that it's safe to use from multiple
concurrent threads. Only one thread can select on a NIO::Selector at a given
time, and while a thread is selecting other threads are blocked from
registering or deregistering IO objects. Once a pending select has completed,
requests to register/unregister IO objects will be processed.

NIO::Selector#wakeup allows one thread to unblock another thread that's in the
middle of an NIO::Selector#select operation. This lets other threads that need
to communicate immediately with the selector unblock it so it can process
other events that it's not presently selecting on.

What nio4r is not
-----------------

nio4r is not a full-featured event framework like EventMachine or Cool.io.
Instead, nio4r is the sort of thing you might write a library like that on
top of. nio4r provides a minimal API such that individual Ruby implementers
may choose to produce optimized versions for their platform, without having
to maintain a large codebase.

As of the time of writing, the current implementation is (approximately):

* 200 lines of Ruby code
* 700 lines of "custom" C code (not counting libev)
* 400 lines of Java code

nio4r is also not a replacement for Kinder Gentler IO (KGIO), a set of
advanced Ruby IO APIs. At some point in the future nio4r might provide a
cross-platform implementation that uses KGIO on CRubies, and Java NIO on JRuby,
however this is not the case today.

License
-------

Copyright (c) 2011-12 Tony Arcieri. Distributed under the MIT License. See
LICENSE.txt for further details.

Includes libev. Copyright (C)2007-09 Marc Alexander Lehmann. Distributed under
the BSD license. See ext/libev/LICENSE for details.


CHANGES.md
HEAD
----
* Merge iobuffer gem in as NIO::Buffer
* OpenSSL::SSL::SSLSocket support

0.3.3
-----
* NIO::Selector#select_each removed
* Remove event buffer
* Patch GIL unlock directly into libev
* Re-release since 0.3.2 was botched :(

0.3.1
-----
* Prevent CancelledKeyExceptions on JRuby

0.3.0
-----
* NIO::Selector#select now takes a block and behaves like select_each
* NIO::Selector#select_each is now deprecated and will be removed
* Closing monitors detaches them from their selector
* Java extension for JRuby
* Upgrade to libev 4.11
* Bugfixes for zero/negative select timeouts
* Handle OP_CONNECT properly on JRuby

0.2.2
-----
* Raise IOError if asked to wake up a closed selector

0.2.1
-----
* Implement wakeup mechanism using raw pipes instead of ev_async, since
  ev_async likes to cause segvs when used across threads (despite claims
  in the documentation to the contrary)

0.2.0
-----
* NIO::Monitor#readiness API to query readiness, along with #readable? and
  #writable? helper methods
* NIO::Selector#select_each API which avoids memory allocations if possible
* Bugfixes for the JRuby implementation

0.1.0
-----
* Initial release. Merry Christmas!


spec/buffer_spec.rb
require File.expand_path(File.dirname(__FILE__) + '/../lib/iobuffer')

describe IO::Buffer do
  before :each do
    @buffer = IO::Buffer.new
    @buffer.size.should == 0
  end

  it "appends data" do
    @buffer.append "foo"
    @buffer.size.should == 3

    @buffer << "bar"
    @buffer.size.should == 6

    @buffer.write "baz"
    @buffer.size.should == 9

    @buffer.read.should == "foobarbaz"
    @buffer.size.should == 0
  end

  it "prepends data" do
    @buffer.prepend "foo"
    @buffer.size.should == 3

    @buffer.prepend "bar"
    @buffer.size.should == 6

    @buffer.read.should == "barfoo"
    @buffer.size.should == 0
  end

  it "mixes prepending and appending properly" do
    source_data = %w{foo bar baz qux}
    actions = permutator([:append, :prepend] * 2)

    actions.each do |sequence|
      sequence.each_with_index do |entry, i|
        @buffer.send(entry, source_data[i])
      end

      @buffer.size.should == sequence.size * 3

      i = 0
      expected = sequence.inject('') do |str, action|
        case action
        when :append
          str << source_data[i]
        when :prepend
          str = source_data[i] + str
        end

        i += 1
        str
      end

      @buffer.read.should == expected
    end
  end

  it "reads data in chunks properly" do
    @buffer.append "foobarbazqux"

    @buffer.read(1).should == 'f'
    @buffer.read(2).should == 'oo'
    @buffer.read(3).should == 'bar'
    @buffer.read(4).should == 'bazq'
    @buffer.read(1).should == 'u'
    @buffer.read(2).should == 'x'
  end

  it "converts to a string" do
    @buffer.append "foobar"
    @buffer.to_str == "foobar"
  end

  it "clears data" do
    @buffer.append "foo"
    @buffer.prepend "bar"

    @buffer.clear
    @buffer.size.should == 0
    @buffer.read.should == ""

    @buffer.prepend "foo"
    @buffer.prepend "bar"
    @buffer.append "baz"

    @buffer.clear
    @buffer.size.should == 0
    @buffer.read.should == ""
  end

  it "knows when it's empty" do
    @buffer.should be_empty
    @buffer.append "foo"
    @buffer.should_not be_empty
  end

  it "can set default node size" do
    IO::Buffer.default_node_size = 1
    IO::Buffer.default_node_size.should == 1
    (IO::Buffer.default_node_size = 4096).should == 4096
    (IO::Buffer.default_node_size = IO::Buffer::MAX_SIZE).should == IO::Buffer::MAX_SIZE
  end

  it "can be created with a different node size" do
    IO::Buffer.new(16384)
  end

  it "cannot set invalid node sizes" do
    proc {
      IO::Buffer.default_node_size = IO::Buffer::MAX_SIZE + 1
    }.should raise_error(ArgumentError)
    proc {
      IO::Buffer.default_node_size = 0
    }.should raise_error(ArgumentError)
    proc {
      IO::Buffer.new(IO::Buffer::MAX_SIZE + 1)
    }.should raise_error(ArgumentError)
    proc {
      IO::Buffer.new(0)
    }.should raise_error(ArgumentError)
  end

  it "Reads can read a single frame" do
    @buffer.append("foo\0bar")
    str = ""
    @buffer.read_frame(str, 0).should == true
    str.should == "foo\0"
    @buffer.size.should == 3
  end

  it "Reads a frame, then reads only some data" do
    @buffer.append("foo\0bar")
    str = ""
    @buffer.read_frame(str,0)
    str = ""
    #This will read only a partial frame
    @buffer.read_frame(str,0).should == false
    str.should == "bar"
    @buffer.size.should == 0
  end

  it "Returns nil when reading from a filehandle at EOF" do
    (rp, wp) = File.pipe

    wp.write("Foo")
    wp.flush
    @buffer.read_from(rp).should == 3
    @buffer.read_from(rp).should == 0
    wp.close
    @buffer.read_from(rp).should == nil
  end

  it "Maintains proper buffer size" do
    #TODO use more methods

    #Testing of normal append
    str = "clarp of the flarn"
    @buffer.append(str)
    s = @buffer.size
    s.should == str.length

    #Testing of read_from
    (rp, wp) = File.pipe
    wp.write(str)
    wp.close
    @buffer.read_from(rp)
    @buffer.size.should == 2*str.length
  end

  it "Can handle lots of data" do
    rp, wp = File.pipe
    srand 1

    total = 0
    100.times do
      chunk_size = rand(2048) #We don't actually know the pipe buffer size!
      if rand > 0.5
        wp.write("x" * chunk_size)
        @buffer.read_from(rp)
      else
        @buffer.append("x" * chunk_size)
      end
      total += chunk_size
    end
    wp.close
    @buffer.read_from(rp)

    @buffer.size.should == total
    @buffer.read.should == "x" * total
  end

  #######
  private
  #######

  def permutator(input)
    output = []
    return output if input.empty?

    (0..input.size - 1).inject([]) do |a, n|
      if a.empty?
        input.each { |x| output << [x] }
      else
        input.each { |x| output += a.map { |y| [x, *y] } }
      end

      output.dup
    end
  end
end


ext/nio4r/buffer.c
/*
 * Copyright (C) 2007-12 Tony Arcieri
 * You may redistribute this under the terms of the MIT license.
 * See LICENSE for details
 */

#include "nio4r.h"
#include <assert.h>

#include <string.h>
#include <time.h>
#include <unistd.h>
#include <errno.h>

/* 1 GiB maximum buffer size */
#define MAX_BUFFER_SIZE 0x40000000

/* Default number of bytes in each node's buffer.  Should be >= MTU */
#define DEFAULT_NODE_SIZE 4096
static unsigned default_node_size = DEFAULT_NODE_SIZE;

struct buffer {
    unsigned size, node_size;
    struct buffer_node *head, *tail;
    struct buffer_node *pool_head, *pool_tail;

};

struct buffer_node {
    unsigned start, end;
    struct buffer_node *next;
    unsigned char   data[0];
};

static VALUE    mNIO = Qnil;
static VALUE    cNIO_Buffer = Qnil;

static VALUE    NIO_Buffer_allocate(VALUE klass);
static void     NIO_Buffer_mark(struct buffer *);
static void     NIO_Buffer_free(struct buffer *);

static VALUE    NIO_Buffer_default_node_size(VALUE klass);
static VALUE    NIO_Buffer_set_default_node_size(VALUE klass, VALUE size);
static VALUE    NIO_Buffer_initialize(int argc, VALUE * argv, VALUE self);
static VALUE    NIO_Buffer_clear(VALUE self);
static VALUE    NIO_Buffer_size(VALUE self);
static VALUE    NIO_Buffer_empty(VALUE self);
static VALUE    NIO_Buffer_append(VALUE self, VALUE data);
static VALUE    NIO_Buffer_prepend(VALUE self, VALUE data);
static VALUE    NIO_Buffer_read(int argc, VALUE * argv, VALUE self);
static VALUE    NIO_Buffer_read_frame(VALUE self, VALUE data, VALUE mark);
static VALUE    NIO_Buffer_to_str(VALUE self);
static VALUE    NIO_Buffer_read_from(VALUE self, VALUE io);
static VALUE    NIO_Buffer_write_to(VALUE self, VALUE io);

static struct buffer *buffer_new(void);
static void     buffer_clear(struct buffer * buf);
static void     buffer_free(struct buffer * buf);
static void     buffer_free_pool(struct buffer * buf);
static void     buffer_prepend(struct buffer * buf, char *str, unsigned len);
static void     buffer_append(struct buffer * buf, char *str, unsigned len);
static void     buffer_read(struct buffer * buf, char *str, unsigned len);
static int      buffer_read_frame(struct buffer * buf, VALUE str, char frame_mark);
static void     buffer_copy(struct buffer * buf, char *str, unsigned len);
static int      buffer_read_from(struct buffer * buf, int fd);
static int      buffer_write_to(struct buffer * buf, int fd);

/*
 * High-performance I/O buffer intended for use in non-blocking programs
 *
 * Data is stored in as a memory-pooled linked list of equally sized chunks.
 * Routines are provided for high speed non-blocking reads and writes from
 * Ruby IO objects.
 */
void
Init_NIO_Buffer()
{
    mNIO = rb_define_module("NIO");
    cNIO_Buffer = rb_define_class_under(mNIO, "Buffer", rb_cObject);
    rb_define_alloc_func(cNIO_Buffer, NIO_Buffer_allocate);

    rb_define_singleton_method(cNIO_Buffer, "default_node_size",
                   NIO_Buffer_default_node_size, 0);
    rb_define_singleton_method(cNIO_Buffer, "default_node_size=",
                   NIO_Buffer_set_default_node_size, 1);

    rb_define_method(cNIO_Buffer, "initialize", NIO_Buffer_initialize, -1);
    rb_define_method(cNIO_Buffer, "clear", NIO_Buffer_clear, 0);
    rb_define_method(cNIO_Buffer, "size", NIO_Buffer_size, 0);
    rb_define_method(cNIO_Buffer, "empty?", NIO_Buffer_empty, 0);
    rb_define_method(cNIO_Buffer, "<<", NIO_Buffer_append, 1);
    rb_define_method(cNIO_Buffer, "append", NIO_Buffer_append, 1);
    rb_define_method(cNIO_Buffer, "write", NIO_Buffer_append, 1);
    rb_define_method(cNIO_Buffer, "prepend", NIO_Buffer_prepend, 1);
    rb_define_method(cNIO_Buffer, "read", NIO_Buffer_read, -1);
    rb_define_method(cNIO_Buffer, "read_frame", NIO_Buffer_read_frame, 2);
    rb_define_method(cNIO_Buffer, "to_str", NIO_Buffer_to_str, 0);
    rb_define_method(cNIO_Buffer, "read_from", NIO_Buffer_read_from, 1);
    rb_define_method(cNIO_Buffer, "write_to", NIO_Buffer_write_to, 1);

    rb_define_const(cNIO_Buffer, "MAX_SIZE", INT2NUM(MAX_BUFFER_SIZE));
}

static VALUE
NIO_Buffer_allocate(VALUE klass)
{
    return Data_Wrap_Struct(klass, NIO_Buffer_mark, NIO_Buffer_free, buffer_new());
}

static void
NIO_Buffer_mark(struct buffer * buf)
{
    /* Naively discard the memory pool whenever Ruby garbage collects */
    buffer_free_pool(buf);
}

static void
NIO_Buffer_free(struct buffer * buf)
{
    buffer_free(buf);
}

/**
 * call-seq:
 *   NIO_Buffer.default_node_size -> 4096
 *
 * Retrieves the current value of the default node size.
 */
static VALUE
NIO_Buffer_default_node_size(VALUE klass)
{
    return UINT2NUM(default_node_size);
}

/*
 * safely converts node sizes from Ruby numerics to C and raising
 * ArgumentError or RangeError on invalid sizes
 */
static unsigned
convert_node_size(VALUE size)
{
    if (
        rb_funcall(size, rb_intern("<"), 1, INT2NUM(1)) == Qtrue ||
        rb_funcall(size, rb_intern(">"), 1, INT2NUM(MAX_BUFFER_SIZE)) == Qtrue
    )
        rb_raise(rb_eArgError, "invalid buffer size");

    return (unsigned) NUM2INT(size);
}

/**
 * call-seq:
 *   NIO_Buffer.default_node_size = 16384
 *
 * Sets the default node size for calling IO::Buffer.new with no arguments.
 */
static VALUE
NIO_Buffer_set_default_node_size(VALUE klass, VALUE size)
{
    default_node_size = convert_node_size(size);

    return size;
}

/**
 *  call-seq:
 *    NIO_Buffer.new(size = IO::Buffer.default_node_size) -> NIO_Buffer
 *
 * Create a new NIO_Buffer with linked segments of the given size
 */
static VALUE
NIO_Buffer_initialize(int argc, VALUE * argv, VALUE self)
{
    VALUE           node_size_obj;
    struct buffer *buf;

    if (rb_scan_args(argc, argv, "01", &node_size_obj) == 1) {
        Data_Get_Struct(self, struct buffer, buf);

        /*
         * Make sure we're not changing the buffer size after data
         * has been allocated
         */
        assert(!buf->head);
        assert(!buf->pool_head);

        buf->node_size = convert_node_size(node_size_obj);
    }
    return Qnil;
}

/**
 *  call-seq:
 *    NIO_Buffer#clear -> nil
 *
 * Clear all data from the NIO_Buffer
 */
static VALUE
NIO_Buffer_clear(VALUE self)
{
    struct buffer *buf;
    Data_Get_Struct(self, struct buffer, buf);

    buffer_clear(buf);

    return Qnil;
}

/**
 *  call-seq:
 *    NIO_Buffer#size -> Integer
 *
 * Return the size of the buffer in bytes
 */
static VALUE
NIO_Buffer_size(VALUE self)
{
    struct buffer *buf;
    Data_Get_Struct(self, struct buffer, buf);

    return INT2NUM(buf->size);
}

/**
 *  call-seq:
 *    NIO_Buffer#empty? -> Boolean
 *
 * Is the buffer empty?
 */
static VALUE
NIO_Buffer_empty(VALUE self)
{
    struct buffer *buf;
    Data_Get_Struct(self, struct buffer, buf);

    return buf->size > 0 ? Qfalse : Qtrue;
}

/**
 *  call-seq:
 *    NIO_Buffer#append(data) -> String
 *
 * Append the given data to the end of the buffer
 */
static VALUE
NIO_Buffer_append(VALUE self, VALUE data)
{
    struct buffer *buf;
    Data_Get_Struct(self, struct buffer, buf);

    /* Is this needed?  Never seen anyone else do it... */
    data = rb_convert_type(data, T_STRING, "String", "to_str");
    buffer_append(buf, RSTRING_PTR(data), RSTRING_LEN(data));

    return data;
}

/**
 *  call-seq:
 *    NIO_Buffer#prepend(data) -> String
 *
 * Prepend the given data to the beginning of the buffer
 */
static VALUE
NIO_Buffer_prepend(VALUE self, VALUE data)
{
    struct buffer *buf;
    Data_Get_Struct(self, struct buffer, buf);

    data = rb_convert_type(data, T_STRING, "String", "to_str");
    buffer_prepend(buf, RSTRING_PTR(data), RSTRING_LEN(data));

    return data;
}

/**
 *  call-seq:
 *    NIO_Buffer#read(length = nil) -> String
 *
 * Read the specified abount of data from the buffer.  If no value
 * is given the entire contents of the buffer are returned.  Any data
 * read from the buffer is cleared.
 * The given length must be greater than 0 or an exception would raise.
 * If the buffer size is zero then an empty string is returned (regardless
 * the given length).
 */
static VALUE
NIO_Buffer_read(int argc, VALUE * argv, VALUE self)
{
    VALUE  length_obj, str;
    int    length;
    struct buffer *buf;

    Data_Get_Struct(self, struct buffer, buf);

    if (rb_scan_args(argc, argv, "01", &length_obj) == 1) {
        length = NUM2INT(length_obj);
        if(length < 1)
          rb_raise(rb_eArgError, "length must be greater than zero");
        if(length > buf->size)
          length = buf->size;
    } else
        length = buf->size;

    if(buf->size == 0)
        return rb_str_new2("");

    str = rb_str_new(0, length);
    buffer_read(buf, RSTRING_PTR(str), length);

    return str;
}

/**
 *  call-seq:
 *    NIO_Buffer#read_frame(str, mark) -> boolean
 *
 * Read up to and including the given frame marker (expressed a a
 * Fixnum 0-255) byte, copying into the supplied string object. If the mark is
 * not encountered before the end of the buffer, false is returned but data
 * is still copied into str. True is returned if the end of a frame is reached.
 *
 */
static VALUE
NIO_Buffer_read_frame(VALUE self, VALUE data, VALUE mark)
{
    char mark_c = (char) NUM2INT(mark);
    struct buffer *buf;

    Data_Get_Struct(self, struct buffer, buf);

    if (buffer_read_frame(buf, data, mark_c)) {
        return Qtrue;
    } else {
        return Qfalse;
    }
}

/**
 *  call-seq:
 *    NIO_Buffer#to_str -> String
 *
 * Convert the Buffer to a String.  The original buffer is unmodified.
 */
static VALUE
NIO_Buffer_to_str(VALUE self)
{
    VALUE          str;
    struct buffer *buf;

    Data_Get_Struct(self, struct buffer, buf);

    str = rb_str_new(0, buf->size);
    buffer_copy(buf, RSTRING_PTR(str), buf->size);

    return str;
}

/**
 *  call-seq:
 *    NIO_Buffer#read_from(io) -> Integer
 *
 * Perform a nonblocking read of the the given IO object and fill
 * the buffer with any data received.  The call will read as much
 * data as it can until the read would block.
 */
static VALUE
NIO_Buffer_read_from(VALUE self, VALUE io)
{
    struct buffer  *buf;
    int             ret;
#if HAVE_RB_IO_T
    rb_io_t        *fptr;
#else
    OpenFile       *fptr;
#endif

    Data_Get_Struct(self, struct buffer, buf);
    GetOpenFile(rb_convert_type(io, T_FILE, "IO", "to_io"), fptr);
    rb_io_set_nonblock(fptr);

    ret = buffer_read_from(buf, FPTR_TO_FD(fptr));
    return ret == -1 ? Qnil : INT2NUM(ret);
}

/**
 *  call-seq:
 *    NIO_Buffer#write_to(io) -> Integer
 *
 * Perform a nonblocking write of the buffer to the given IO object.
 * As much data as possible is written until the call would block.
 * Any data which is written is removed from the buffer.
 */
static VALUE
NIO_Buffer_write_to(VALUE self, VALUE io)
{
    struct buffer  *buf;
#if HAVE_RB_IO_T
    rb_io_t        *fptr;
#else
    OpenFile       *fptr;
#endif

    Data_Get_Struct(self, struct buffer, buf);
    GetOpenFile(rb_convert_type(io, T_FILE, "IO", "to_io"), fptr);
    rb_io_set_nonblock(fptr);

    return INT2NUM(buffer_write_to(buf, FPTR_TO_FD(fptr)));
}

/*
 * Ruby bindings end here.  Below is the actual implementation of
 * the underlying byte queue ADT
 */

/* Create a new buffer */
static struct buffer *
buffer_new(void)
{
    struct buffer *buf;

    buf = (struct buffer *) xmalloc(sizeof(struct buffer));
    buf->head = buf->tail = buf->pool_head = buf->pool_tail = 0;
    buf->size = 0;
    buf->node_size = default_node_size;

    return buf;
}

/* Clear all data from a buffer */
static void
buffer_clear(struct buffer * buf)
{
    /* Move everything into the buffer pool */
    if (!buf->pool_tail) {
        buf->pool_head = buf->pool_tail = buf->head;
    } else {
        buf->pool_tail->next = buf->head;
    }

    buf->head = buf->tail = 0;
    buf->size = 0;
}

/* Free a buffer */
static void
buffer_free(struct buffer * buf)
{
    buffer_clear(buf);
    buffer_free_pool(buf);

    free(buf);
}

/* Free the memory pool */
static void
buffer_free_pool(struct buffer * buf)
{
    struct buffer_node *tmp;

    while (buf->pool_head) {
        tmp = buf->pool_head;
        buf->pool_head = tmp->next;
        free(tmp);
    }

    buf->pool_tail = 0;
}

/* Create a new buffer_node (or pull one from the memory pool) */
static struct buffer_node *
buffer_node_new(struct buffer * buf)
{
    struct buffer_node *node;

    /* Pull from the memory pool if available */
    if (buf->pool_head) {
        node = buf->pool_head;
        buf->pool_head = node->next;

        if (node->next)
            node->next = 0;
        else
            buf->pool_tail = 0;
    } else {
        node = (struct buffer_node *) xmalloc(sizeof(struct buffer_node) + buf->node_size);
        node->next = 0;
    }

    node->start = node->end = 0;
    return node;
}

/* Free a buffer node (i.e. return it to the memory pool) */
static void
buffer_node_free(struct buffer * buf, struct buffer_node * node)
{
    node->next = buf->pool_head;
    buf->pool_head = node;

    if (!buf->pool_tail) {
        buf->pool_tail = node;
    }
}

/* Prepend data to the front of the buffer */
static void
buffer_prepend(struct buffer * buf, char *str, unsigned len)
{
    struct buffer_node *node, *tmp;
    buf->size += len;

    /* If it fits in the beginning of the head */
    if (buf->head && buf->head->start >= len) {
        buf->head->start -= len;
        memcpy(buf->head->data + buf->head->start, str, len);
    } else {
        node = buffer_node_new(buf);
        node->next = buf->head;
        buf->head = node;
        if (!buf->tail)
            buf->tail = node;

        while (len > buf->node_size) {
            memcpy(node->data, str, buf->node_size);
            node->end = buf->node_size;

            tmp = buffer_node_new(buf);
            tmp->next = node->next;
            node->next = tmp;

            if (buf->tail == node)
                buf->tail = tmp;
            node = tmp;

            str += buf->node_size;
            len -= buf->node_size;
        }

        if (len > 0) {
            memcpy(node->data, str, len);
            node->end = len;
        }
    }
}

/* Append data to the front of the buffer */
static void
buffer_append(struct buffer * buf, char *str, unsigned len)
{
    unsigned nbytes;
    buf->size += len;

    /* If it fits in the remaining space in the tail */
    if (buf->tail && len <= buf->node_size - buf->tail->end) {
        memcpy(buf->tail->data + buf->tail->end, str, len);
        buf->tail->end += len;
        return;
    }
    /* Empty list needs initialized */
    if (!buf->head) {
        buf->head = buffer_node_new(buf);
        buf->tail = buf->head;
    }
    /* Build links out of the data */
    while (len > 0) {
        nbytes = buf->node_size - buf->tail->end;
        if (len < nbytes)
            nbytes = len;

        memcpy(buf->tail->data + buf->tail->end, str, nbytes);
        str += nbytes;
        len -= nbytes;

        buf->tail->end += nbytes;

        if (len > 0) {
            buf->tail->next = buffer_node_new(buf);
            buf->tail = buf->tail->next;
        }
    }
}

/* Read data from the buffer (and clear what we've read) */
static void
buffer_read(struct buffer * buf, char *str, unsigned len)
{
    unsigned nbytes;
    struct buffer_node *tmp;

    while (buf->size > 0 && len > 0) {
        nbytes = buf->head->end - buf->head->start;
        if (len < nbytes)
            nbytes = len;

        memcpy(str, buf->head->data + buf->head->start, nbytes);
        str += nbytes;
        len -= nbytes;

        buf->head->start += nbytes;
        buf->size -= nbytes;

        if (buf->head->start == buf->head->end) {
            tmp = buf->head;
            buf->head = tmp->next;
            buffer_node_free(buf, tmp);

            if (!buf->head)
                buf->tail = 0;
        }
    }
}

/*
 * Read data from the buffer into str until byte frame_mark or empty. Bytes
 * are copied into str and removed if a complete frame is read, a true value
 * is returned
 */
static int
buffer_read_frame(struct buffer * buf, VALUE str, char frame_mark)
{
    unsigned nbytes = 0;
    struct buffer_node *tmp;

    while (buf->size > 0) {
        struct buffer_node *head = buf->head;
        char           *loc, *s = head->data + head->start, *e = head->data + head->end;
        nbytes = e - s;

        loc = memchr(s, frame_mark, nbytes);

        if (loc) {
            nbytes = loc - s + 1;
        }

        /* Copy less than everything if we found a frame byte */
        rb_str_cat(str, s, nbytes);

        /* Fixup the buffer pointers to indicate the bytes were consumed */
        head->start += nbytes;
        buf->size -= nbytes;

        if (head->start == head->end) {
            buf->head = head->next;
            buffer_node_free(buf, head);

            if (!buf->head)
                buf->tail = 0;
        }

        if (loc) {
            return 1;
        }
    }

    return 0;
}

/* Copy data from the buffer without clearing it */
static void
buffer_copy(struct buffer * buf, char *str, unsigned len)
{
    unsigned nbytes;
    struct buffer_node *node;

    node = buf->head;
    while (node && len > 0) {
        nbytes = node->end - node->start;
        if (len < nbytes)
            nbytes = len;

        memcpy(str, node->data + node->start, nbytes);
        str += nbytes;
        len -= nbytes;

        if (node->start + nbytes == node->end)
            node = node->next;
    }
}

/* Write data from the buffer to a file descriptor */
static int
buffer_write_to(struct buffer * buf, int fd)
{
    int bytes_written, total_bytes_written = 0;
    struct buffer_node *tmp;

    while (buf->head) {
        bytes_written = write(fd, buf->head->data + buf->head->start, buf->head->end - buf->head->start);

        /* If the write failed... */
        if (bytes_written < 0) {
            if (errno != EAGAIN)
                rb_sys_fail("write");

            return total_bytes_written;
        }

        total_bytes_written += bytes_written;
        buf->size -= bytes_written;

        /* If the write blocked... */
        if (bytes_written < buf->head->end - buf->head->start) {
            buf->head->start += bytes_written;
            return total_bytes_written;
        }
        /* Otherwise we wrote the whole buffer */
        tmp = buf->head;
        buf->head = tmp->next;
        buffer_node_free(buf, tmp);

        if (!buf->head)
            buf->tail = 0;
    }

    return total_bytes_written;
}

/* Read data from a file descriptor to a buffer */
/* Append data to the front of the buffer */
static int
buffer_read_from(struct buffer * buf, int fd)
{
    int      bytes_read, total_bytes_read = 0;
    unsigned nbytes;

    /* Empty list needs initialized */
    if (!buf->head) {
        buf->head = buffer_node_new(buf);
        buf->tail = buf->head;
    }

    do {
        nbytes = buf->node_size - buf->tail->end;
        bytes_read = read(fd, buf->tail->data + buf->tail->end, nbytes);

        if (bytes_read == 0) {
            return -1;
            //When the file reaches EOF
        } else if (bytes_read < 0) {
            if (errno != EAGAIN)
                rb_sys_fail("read");

            return total_bytes_read;
        }

        total_bytes_read += bytes_read;
        buf->tail->end += bytes_read;
        buf->size += bytes_read;

        if (buf->tail->end == buf->node_size) {
            buf->tail->next = buffer_node_new(buf);
            buf->tail = buf->tail->next;
        }
    } while (bytes_read == nbytes);

    return total_bytes_read;
}
